File: android/app/src/main/java/com/elderlyassistant/services/medication/DoubleDoseDetector.kt
package com.elderlyassistant.services.medication

import java.util.UUID

// MARK: - Double Dose Detector (FR-D03)

data class DoubleDoseCheckResult(
    val isDuplicate: Boolean,
    val previousDoseAt: Long?,
    val windowEndsAt: Long?
)

class DoubleDoseDetector {
    fun check(
        medicationEntryId: UUID,
        windowHours: Int,
        adherenceLog: List<MedicationAdherenceLog>,
        now: Long
    ): DoubleDoseCheckResult {
        val windowStart = now - (windowHours * 3600_000L)

        val previousDose = adherenceLog
            .filter { it.medicationEntryId == medicationEntryId }
            .filter { it.status == MedicationAdherenceStatus.ACKNOWLEDGED }
            .filter { (it.acknowledgedAt ?: 0) in windowStart..now }
            .maxByOrNull { it.acknowledgedAt ?: 0 }

        if (previousDose == null) {
            return DoubleDoseCheckResult(false, null, null)
        }

        val previousAt = previousDose.acknowledgedAt!!
        val windowEnds = previousAt + (windowHours * 3600_000L)

        return DoubleDoseCheckResult(true, previousAt, windowEnds)
    }
}

File: ios/ElderlyAssistant/Services/MedicationScheduler/DoubleDoseDetector.swift
import Foundation

// MARK: - Double Dose Detector (FR-D03)

struct DoubleDoseCheckResult {
    let isDuplicate: Bool
    let previousDoseAt: Date?
    let windowEndsAt: Date?
}

final class DoubleDoseDetector {
    /// Check whether the given medication has already been acknowledged within its
    /// double-dose prevention window.
    ///
    /// - Parameters:
    ///   - medicationEntryId: The medication being checked
    ///   - windowHours: Configurable window (default 4 for once-daily, 2 for twice-daily)
    ///   - adherenceLog: The on-device adherence log entries for this medication today
    ///   - now: Current time
    /// - Returns: DoubleDoseCheckResult indicating whether a duplicate was found
    func check(
        medicationEntryId: UUID,
        windowHours: Int,
        adherenceLog: [MedicationAdherenceLog],
        now: Date
    ) -> DoubleDoseCheckResult {
        let windowStart = now.addingTimeInterval(-TimeInterval(windowHours * 3600))

        let previousDose = adherenceLog
            .filter { $0.medicationEntryId == medicationEntryId }
            .filter { $0.status == .acknowledged }
            .filter { log in
                if let ackAt = log.acknowledgedAt {
                    return ackAt >= windowStart && ackAt <= now
                }
                return false
            }
            .sorted { a, b in
                (a.acknowledgedAt ?? .distantPast) > (b.acknowledgedAt ?? .distantPast)
            }
            .first

        guard let previous = previousDose, let previousAt = previous.acknowledgedAt else {
            return DoubleDoseCheckResult(
                isDuplicate: false,
                previousDoseAt: nil,
                windowEndsAt: nil
            )
        }

        let windowEndsAt = previousAt.addingTimeInterval(TimeInterval(windowHours * 3600))

        return DoubleDoseCheckResult(
            isDuplicate: true,
            previousDoseAt: previousAt,
            windowEndsAt: windowEndsAt
        )
    }
}
